pkg/apis/resource/validation/validation.go

```diff
@@ -693,3 +708,71 @@ func validateMap[K ~string, T any](m map[K]T, maxSize int, validateKey func(K, *
 	}
 	return allErrs
 }
+
+func validateDeviceStatus(device resource.AllocatedDeviceStatus, fldPath *field.Path, allocatedDevices sets.Set[structured.DeviceID]) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateDriverName(device.Driver, fldPath.Child("driver"))...)
+	allErrs = append(allErrs, validatePoolName(device.Pool, fldPath.Child("pool"))...)
+	allErrs = append(allErrs, validateDeviceName(device.Device, fldPath.Child("device"))...)
+	deviceID := structured.DeviceID{Driver: device.Driver, Pool: device.Pool, Device: device.Device}
+	if !allocatedDevices.Has(deviceID) {
+		allErrs = append(allErrs, field.Invalid(fldPath, deviceID, "must be an allocated device in the claim"))
+	}
+	allErrs = append(allErrs, metav1validation.ValidateConditions(device.Conditions, fldPath.Child("conditions"))...)
+	if len(device.Data.Raw) > 0 { // Data is an optional field.
+		allErrs = append(allErrs, validateRawExtension(device.Data, fldPath.Child("data"))...)
+	}
+	allErrs = append(allErrs, validateNetworkDeviceData(device.NetworkData, fldPath.Child("networkData"))...)
+	return allErrs
+}
+
+// validateRawExtension validates RawExtension as in https://github.com/kubernetes/kubernetes/pull/125549/
+func validateRawExtension(rawExtension runtime.RawExtension, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	var v any
+	if len(rawExtension.Raw) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else if err := json.Unmarshal(rawExtension.Raw, &v); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, "<value omitted>", fmt.Sprintf("error parsing data as JSON: %v", err.Error())))
+	} else if v == nil {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else if _, isObject := v.(map[string]any); !isObject {
+		allErrs = append(allErrs, field.Invalid(fldPath, "<value omitted>", "parameters must be a valid JSON object"))
+	}
+	return allErrs
+}
+
+const interfaceNameMaxLength int = 256
+const hardwareAddressMaxLength int = 128
+
+func validateNetworkDeviceData(networkDeviceData *resource.NetworkDeviceData, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if networkDeviceData == nil {
+		return allErrs
+	}
+
+	if len(networkDeviceData.InterfaceName) > interfaceNameMaxLength {
+		allErrs = append(allErrs, field.TooLongMaxLength(fldPath.Child("interfaceName"), networkDeviceData.InterfaceName, interfaceNameMaxLength))
```

Comment:
TooLongMaxLength is going away - please use `field.TooLong(fldPath.Child("interfaceName"), "" /* unused */, interfaceNameMaxLength)`

NOTE: the "value" arg is unused